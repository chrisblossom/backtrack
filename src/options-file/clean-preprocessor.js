/* @flow */

import path from 'path';
import { toArray } from '../utils/object-utils';
import { buildPath, rootPath } from '../config/paths';
import type { Clean, NormalizedClean } from '../types.js';

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value?: Clean | $ReadOnlyArray<Clean>,
};

function cleanPreprocessor({ value }: Args = {}): $ReadOnlyArray<
    NormalizedClean,
> {
    const allToArray = toArray(value);

    return allToArray.map((arg: Clean) => {
        const makeDirs = toArray(arg.makeDirs).map((makeDir) => {
            const resolvedMakeDir = path.resolve(buildPath, makeDir);
            return resolvedMakeDir;
        });

        const copy = toArray(arg.copy).map((copyArgs) => {
            const resolvedSrc = path.resolve(rootPath, copyArgs.src);
            const resolvedDest = path.resolve(buildPath, copyArgs.dest);

            const result = {
                src: resolvedSrc,
                dest: resolvedDest,
                hash: copyArgs.hash,
            };

            if (result.hash === undefined) {
                delete result.hash;
            }

            return result;
        });

        return {
            del: toArray(arg.del),
            makeDirs,
            copy,
        };
    });
}

export { cleanPreprocessor };
