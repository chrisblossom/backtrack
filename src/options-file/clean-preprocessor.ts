import path from 'path';
import { toArray } from '../utils/object-utils';
import { buildPath, rootPath } from '../config/paths';
import { Clean, NormalizedClean } from '../types';

interface Args {
	value?: Clean | Clean[];
}

function cleanPreprocessor({ value }: Args = {}): NormalizedClean[] {
	const allToArray = toArray(value);

	return allToArray.map((arg: Clean) => {
		const makeDirs = toArray(arg.makeDirs).map((makeDir) => {
			const resolvedMakeDir = path.resolve(buildPath, makeDir);
			return resolvedMakeDir;
		});

		const copy = toArray(arg.copy).map((copyArgs) => {
			const resolvedSrc = path.resolve(rootPath, copyArgs.src);
			const resolvedDest = path.resolve(buildPath, copyArgs.dest);

			const result = {
				src: resolvedSrc,
				dest: resolvedDest,
				hash: copyArgs.hash,
			};

			if (result.hash === undefined) {
				delete result.hash;
			}

			return result;
		});

		return {
			del: toArray(arg.del),
			makeDirs,
			copy,
		};
	});
}

export { cleanPreprocessor };
