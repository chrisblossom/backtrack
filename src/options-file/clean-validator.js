/* @flow */

import { buildPath } from '../config/paths';
import { fileIsInsideDir } from '../utils/file-is-inside-dir';

import type { NormalizedClean } from '../types.js';

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value: $ReadOnlyArray<NormalizedClean>,
};

function cleanValidator({ value }: Args) {
    for (const arg of value) {
        for (const pattern of arg.del) {
            if (fileIsInsideDir(pattern, buildPath) === false) {
                throw new Error(
                    `del pattern '${pattern}' must be inside build directory`,
                );
            }
        }

        for (const dir of arg.makeDirs) {
            if (fileIsInsideDir(dir, buildPath) === false) {
                throw new Error(
                    `makeDirs '${dir}' must be inside build directory`,
                );
            }
        }
    }
}

export { cleanValidator };
