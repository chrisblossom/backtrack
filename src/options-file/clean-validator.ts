import path from 'path';
import { buildPath, rootPath } from '../config/paths';
import { fileIsInsideDirSync } from '../utils/file-is-inside-dir';
import { NormalizedClean } from '../types';

interface Args {
	value: NormalizedClean[];
}

function cleanValidator({ value }: Args): void {
	for (const arg of value) {
		for (const pattern of arg.del) {
			const resolvedPattern = path.resolve(buildPath, pattern);
			if (fileIsInsideDirSync(resolvedPattern, buildPath) === false) {
				throw new Error(
					`del pattern '${pattern}' must be inside build directory`,
				);
			}
		}

		for (const dir of arg.makeDirs) {
			if (fileIsInsideDirSync(dir, buildPath) === false) {
				throw new Error(
					`makeDirs '${dir}' must be inside build directory`,
				);
			}
		}

		for (const copy of arg.copy) {
			if (fileIsInsideDirSync(copy.src, rootPath) === false) {
				throw new Error(
					`copy.src '${copy.src}' must be inside project root directory`,
				);
			}

			if (fileIsInsideDirSync(copy.src, buildPath) === true) {
				throw new Error(
					`copy.src '${copy.src}' cannot be inside build directory`,
				);
			}

			if (fileIsInsideDirSync(copy.dest, buildPath) === false) {
				throw new Error(
					`copy.dest '${copy.dest}' must be inside build directory`,
				);
			}

			const isDuplicateWithHashMismatch = !!arg.copy.find((copyArg) => {
				return (
					copyArg.src === copy.src &&
					copyArg.dest === copy.dest &&
					copyArg.hash !== copy.hash
				);
			});

			if (isDuplicateWithHashMismatch) {
				throw new Error(
					`copy.src '${copy.src}' is duplicated with mismatched hash: true/false`,
				);
			}

			const isDuplicateWithSrcMismatch = !!arg.copy.find((copyArg) => {
				return copyArg.src !== copy.src && copyArg.dest === copy.dest;
			});

			if (isDuplicateWithSrcMismatch) {
				throw new Error(
					`copy.dest '${copy.dest}' is duplicated with mismatched sources`,
				);
			}
		}
	}
}

export { cleanValidator };
