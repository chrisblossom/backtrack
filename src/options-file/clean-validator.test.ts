import path from 'path';
import os from 'os';

const cleanValidator = (args: any) =>
	require('./clean-validator').cleanValidator(args);

const cleanPreprocessor = (args: any) =>
	require('./clean-preprocessor').cleanPreprocessor(args);

function getValue(value: any) {
	return cleanPreprocessor({ value });
}

describe('cleanValidator', () => {
	const cwd = process.cwd();
	const dir = path.resolve(__dirname, '../clean/__sandbox__/1/');

	beforeEach(() => {
		process.chdir(dir);
	});

	afterEach(() => {
		process.chdir(cwd);
	});

	test('del fails with relative ../', () => {
		const value = getValue([
			{
				del: ['../src/**'],
				makeDirs: [],
				copy: [],
			},
		]);

		let error;
		try {
			cleanValidator({ value });
		} catch (e) {
			error = e;
		} finally {
			expect(error).toMatchSnapshot();
		}
	});

	test('del fails with absolute inside rootPath', () => {
		const value = getValue([
			{
				del: [path.resolve(dir, 'src/**')],
				makeDirs: [],
				copy: [],
			},
		]);

		let error;
		try {
			cleanValidator({ value });
		} catch (e) {
			error = e;
		} finally {
			expect(error).toMatchSnapshot();
		}
	});

	test('del fails with absolute', () => {
		process.chdir(__dirname);
		const value = getValue([
			{
				del: [path.resolve(__dirname, '__sandbox__/should_fail')],
				makeDirs: [],
				copy: [],
			},
		]);

		let error;
		try {
			cleanValidator({ value });
		} catch (e) {
			error = e;
		} finally {
			expect(error).toMatchSnapshot();
		}
	});

	test('del fails with absolute + relative', () => {
		process.chdir(__dirname);
		const value = getValue([
			{
				del: [path.resolve(__dirname, '__sandbox__/1/../dist/')],
				makeDirs: [],
				copy: [],
			},
		]);

		let error;
		try {
			cleanValidator({ value });
		} catch (e) {
			error = e;
		} finally {
			expect(error).toMatchSnapshot();
		}
	});

	test('makeDirs fails when outside build directory', () => {
		process.chdir(__dirname);
		const value = getValue([
			{
				del: [],
				makeDirs: [path.resolve(__dirname, '__sandbox__/1/../dist/')],
				copy: [],
			},
		]);

		let error;
		try {
			cleanValidator({ value });
		} catch (e) {
			error = e;
		} finally {
			expect(error).toMatchSnapshot();
		}
	});

	test('copy fails when dest outside build directory', () => {
		process.chdir(dir);
		const value = getValue([
			{
				del: [],
				makeDirs: [],
				copy: [
					{
						dest: path.resolve(dir, 'static'),
						src: path.resolve(dir, 'static'),
					},
				],
			},
		]);

		let error;
		try {
			cleanValidator({ value });
		} catch (e) {
			error = e;
		} finally {
			expect(error).toMatchSnapshot();
		}
	});

	test('copy fails when src outside root directory', () => {
		process.chdir(dir);
		const value = getValue([
			{
				del: [],
				makeDirs: [],
				copy: [
					{
						dest: path.resolve(dir, 'dist/static'),
						src: path.resolve(os.homedir(), 'static'),
					},
				],
			},
		]);

		let error;
		try {
			cleanValidator({ value });
		} catch (e) {
			error = e;
		} finally {
			expect(error).toMatchSnapshot();
		}
	});

	test('copy fails when src is inside build directory', () => {
		process.chdir(dir);
		const value = getValue([
			{
				del: [],
				makeDirs: [],
				copy: [
					{
						dest: path.resolve(dir, 'dist/static'),
						src: path.resolve(dir, 'dist/static'),
					},
				],
			},
		]);

		let error;
		try {
			cleanValidator({ value });
		} catch (e) {
			error = e;
		} finally {
			expect(error).toMatchSnapshot();
		}
	});

	test('copy fails when duplicate dest but differing src', () => {
		process.chdir(dir);
		const value = getValue([
			{
				del: [],
				makeDirs: [],
				copy: [
					{
						src: 'static-1',
						dest: 'static',
					},
					{
						src: 'static-2',
						dest: 'static',
					},
				],
			},
		]);

		let error;
		try {
			cleanValidator({ value });
		} catch (e) {
			error = e;
		} finally {
			expect(error).toMatchSnapshot();
		}
	});

	test('copy fails when duplicate dest/src and mismatched hash: true/false', () => {
		process.chdir(dir);
		const value = getValue([
			{
				del: [],
				makeDirs: [],
				copy: [
					{
						dest: 'static',
						src: 'static',
						hash: true,
					},
					{
						dest: 'static',
						src: 'static',
						hash: false,
					},
				],
			},
		]);

		let error;
		try {
			cleanValidator({ value });
		} catch (e) {
			error = e;
		} finally {
			expect(error).toMatchSnapshot();
		}
	});

	test('passes with just del', () => {
		const value = getValue([
			{
				del: ['*'],
				makeDirs: [],
				copy: [],
			},
		]);

		const result = cleanValidator({ value });
		expect(result).toEqual(undefined);
	});

	test('passes with del glob', () => {
		const value = getValue([
			{
				del: ['!.git'],
				makeDirs: [],
				copy: [],
			},
		]);

		const result = cleanValidator({ value });
		expect(result).toEqual(undefined);
	});

	test('passes with just makeDirs', () => {
		const value = getValue([
			{
				del: [],
				makeDirs: ['fake/dir/'],
				copy: [],
			},
		]);

		const result = cleanValidator({ value });
		expect(result).toEqual(undefined);
	});

	test('passes with just copy', () => {
		process.chdir(dir);
		const value = getValue([
			{
				del: [],
				makeDirs: [],
				copy: [
					{
						dest: 'static',
						src: 'static',
					},
				],
			},
		]);

		const result = cleanValidator({ value });
		expect(result).toEqual(undefined);
	});
});
