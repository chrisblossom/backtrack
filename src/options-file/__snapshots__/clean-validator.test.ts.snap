// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cleanValidator copy fails when dest outside build directory 1`] = `[Error: copy.dest '<PROJECT_ROOT>/static' must be inside build directory]`;

exports[`cleanValidator copy fails when duplicate dest but differing src 1`] = `[Error: copy.dest '<PROJECT_ROOT>/dist/static' is duplicated with mismatched sources]`;

exports[`cleanValidator copy fails when duplicate dest/src and mismatched hash: true/false 1`] = `[Error: copy.src '<PROJECT_ROOT>/static' is duplicated with mismatched hash: true/false]`;

exports[`cleanValidator copy fails when src is inside build directory 1`] = `[Error: copy.src '<PROJECT_ROOT>/dist/static' cannot be inside build directory]`;

exports[`cleanValidator copy fails when src outside root directory 1`] = `[Error: copy.src '<HOME_DIR>/static' must be inside project root directory]`;

exports[`cleanValidator del fails with absolute + relative 1`] = `[Error: del pattern '<PROJECT_ROOT>/__sandbox__/dist' must be inside build directory]`;

exports[`cleanValidator del fails with absolute 1`] = `[Error: del pattern '<PROJECT_ROOT>/__sandbox__/should_fail' must be inside build directory]`;

exports[`cleanValidator del fails with absolute inside rootPath 1`] = `[Error: del pattern '<PROJECT_ROOT>/src/**' must be inside build directory]`;

exports[`cleanValidator del fails with relative ../ 1`] = `[Error: del pattern '../src/**' must be inside build directory]`;

exports[`cleanValidator makeDirs fails when outside build directory 1`] = `[Error: makeDirs '<PROJECT_ROOT>/__sandbox__/dist' must be inside build directory]`;
